### DISCUSSION TITLE
Merging spans after creating custom attributes

### DISCUSSION THREAD

**Discussion by Fourthought** [explosion/spaCy#4806] [Help: Coding & Implementations]
I've been using the code examples from the documentation, particularly the 'Custom pipeline components and attribute extensions via a REST API' example. After creating the custom attribute extension there is a section of the code for creating the new spans as follows:

```
for span in spans:
    # Iterate over all spans and merge them into one token. This is done
    # after setting the entities – otherwise, it would cause mismatched
    # indices!
    span.merge()
```

I see, however, that `merge()` has been deprecated, so have developed the following (rather inelegant) solution for the same task:

```
for span in spacy.util.filter_spans(spans):     ## span is a list of spans with custom attribute
    with doc.retokenize() as retokenizer:
        retokenizer.merge(span)                 ## merge the spans

        doc._.named_concepts = spacy.util.filter_spans(list(doc._.named_concepts) + [span])     ## filter the list of `docs.ents` and new spans to resolve any overlapping entities
```
`spans` is a list object containing the spans derived from the `matcher`

This code is throwing up the following error message:

`IndexError: [E036] Error calculating span: Can't find a token starting at character offset 1992.`

Have also tried this code:

```
doc._.named_concepts = spacy.util.filter_spans(list(doc._.named_concepts) + spans) # concept span
        
        for span in spacy.util.filter_spans(spans):
            with doc.retokenize() as retokenizer:
                retokenizer.merge(span)
```
...but have got the same result.

Is there a more elegant solution to allow the creation of custom extensions please?

I would use the Entity Ruler, but am looking to modify the spans as they are created.

## Which page or section is this issue related to?
[(https://spacy.io/usage/examples)](url)

**Comment by adrianeboyd** [comment]
Notice that the example script sets all the entities before doing any merging. Otherwise the saved spans from the `Matcher` won't have the right token indices anymore because they refer to the original tokenization, not the new merged (and renumbered) tokens.

The update for the example script would be:

Current version (note the warning in the comment):

```
        for span in spans:
            # Iterate over all spans and merge them into one token. This is done
            # after setting the entities – otherwise, it would cause mismatched
            # indices!
            span.merge()
```

Updated version:

```
        with doc.retokenize() as retokenizer:
            # Iterate over all spans and merge them into one token. This is done
            # after setting the entities – otherwise, it would cause mismatched
            # indices!
            for span in spans:
                retokenizer.merge(span)
```

**Comment by Fourthought** [comment]
Hi Adriane,

Thank you for your response, however, I still seem to be having problems. Having followed your advice the indices are incorrect. `token._.concept` returns the correct value, however, `doc[token.start]._.CONCEPT` is not working.

So i've tried to work through the problem and have tried numerous permutations, the lastest version is here:

`matches = self.matcher(doc)
        spans = []  # keep the spans for later so we can merge them afterwards
        
        for match_id, start, end in matches:
            
            concept = Span(doc, start, end)
            
            ## gather up noun phrases 
            ## I found the subtree attribute to be a bit too greedy on some key tokens. 
            ## Have also tried without this section, but got the same results
            
            if doc[start - 1].dep_ in ["amod", "compound"] and start != 0: 
                concept = Span(doc, start - 1, end)
                
            elif doc[start - 1].pos_ in ["NOUN", "PROPN"] and start != 0:
                concept = Span(doc, start - 1, end)
            
            elif doc[start - 2].dep_ in ["nsubj", "amod"] and doc[start].dep_ in ["pobj"] and start != 0:
                concept = Span(doc, start - 2, end)
                
            elif doc[start].dep_ in ["nsubj", "csubj", "pobj"] and end != len(doc):
                if doc[start + 1].dep_ in ["prep"]: 
                    try:
                        concept = Span(doc, start, list(doc[start + 1].rights)[-1].i + 1)
                    except:
                        continue            
                        
            ## get labels for the concept
            concept._.CONCEPT = doc.vocab.strings[match_id]
            concept._.INFLUENCE = Orator().get_influence(concept._.CONCEPT)
            
            spans.append(concept)

        doc._.named_concepts = spacy.util.filter_spans(list(doc._.named_concepts) + spans)
        
        with doc.retokenize() as retokenizer:
            for span in doc._.named_concepts:
                retokenizer.merge(span)`

This version, however, is returning the following error when iterating over the `Doc._.named_concepts` extension:
`IndexError: [E036] Error calculating span: Can't find a token starting at character offset 1246.`

Do you know where I'm going wrong here please?

I know the answer is going to be a simple one, but I just can't see it...

**Comment by adrianeboyd** [comment]
Ah, sorry, I missed the actual problem the first time around: you're adding the old spans to `._.named_concepts`, so those spans refers to the old token offsets rather than the new ones:

```
doc._.named_concepts = spacy.util.filter_spans(list(doc._.named_concepts) + [span])
```

This is tricky when you're merging because the span offsets will constantly shift, so it's better not to try to maintain a custom `Doc` attribute with the spans but to keep the information at the token level, at least until you're done merging. (`doc.ents` is easy to adjust during retokenization because the entities are stored underneath as IOB tags on the token level.)

Let's see, the main difference to the deprecated `span.merge()` method is that the retokenizer doesn't return a reference to the modified tokens, so you need a way to recover the merged tokens afterwards. You might be able to just use your span extensions to identify the concepts after merging:

```
for token in doc:
    print(token.text, doc[token.i:token.i+1]._.CONCEPT)
```

But if there's some reason that that wouldn't work because of how you're processing the documents, another way is to add a custom attribute as you're merging and then use that to identify the merged tokens afterwards. Here's an example with custom span and token attributes that get merged:

```
import spacy
from spacy.matcher import Matcher
from spacy.tokens import Token, Span
from spacy.util import filter_spans

nlp = spacy.load('en')
matcher = Matcher(nlp.vocab)
matcher.add("CONCEPT", None, [{"LOWER": "my"}, {"LOWER": "concept"}])
text = "My concept is mentioned here. This is my concept."
doc = nlp(text)

Token.set_extension("merged_concept", default=False)
Span.set_extension("concept_id", default=0)

matches = matcher(doc)
spans = [Span(doc, match[1], match[2]) for match in matches]
# just to show that the retokenizer also handles custom span extensions
for i, span in enumerate(spans):
    span._.concept_id = i + 1

with doc.retokenize() as retokenizer:
    for span in filter_spans(spans):
        retokenizer.merge(span, attrs={"_": {"merged_concept": True}})

for token in doc:
    print(token.text, token._.merged_concept, doc[token.i:token.i+1]._.concept_id)
```

Output:

```
My concept True 1
is False 0
mentioned False 0
here False 0
. False 0
This False 0
is False 0
my concept True 2
. False 0
```

**Comment by Fourthought** [comment]
Adriane,

This is great, thank you for your comprehensive response. Let me take a look and I'll get back to you with the results.

Have a great Christmas!

Steve
